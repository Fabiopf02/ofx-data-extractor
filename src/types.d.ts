export type OFXMetaData = {
  OFXHEADER: string
  DATA: string
  VERSION: string
  SECURITY: string
  ENCODING: string
  CHARSET: string
  COMPRESSION: string
  OLDFILEUID: string
  NEWFILEUID: string
}

export type DateResponse = {
  datetime: string | null
  date: string | null
  time: string | null
  offset: string | null
  timezone: string | null
}

/**
 * separated - returns the separate transaction id: 
 *  `{
 *    "date": string,
      "transactionCode": string,
      "protocol": string
    }`
 */
type ConfigFitId = 'normal' | 'separated'

export type ConfigDate = {
  /**
   * @description supported keys:
   *  yy => year -> 2 digits,
   *  yyyy or y => year,
   *  MM or M => month,
   *  dd or d => day,
   *  hh or h => hour,
   *  mm or m => minute,
   *  ss or s => second,
   *  O => offset,
   *  TZ => timezone
   * @example format: 'y-M-d h:m:s'
   * @returns '2022-02-21 09:00:00'
   */
  formatDate?: FormatDate
}

type MapKeys<T> = {
  [Property in keyof T]?: string
}

type KeysJson = MapKeys<
  | Status
  | BankAccountFrom
  | STRTTRN
  | BankTransferList
  | LedGerBal
  | FINANCIAL_INSTITUTION
  | OfxStructure
>

export type OfxConfig = ConfigDate & {
  fitId?: ConfigFitId
  nativeTypes?: boolean
  formatJson?: MapKeys<KeysJson>
}

type Status = {
  CODE: string
  SEVERITY: string
}

type BankAccountFrom = {
  BANKID: string
  ACCTID: string
  ACCTT: string
}

type TransferType = string

type STRTTRN = {
  TRNTYPE: TransferType
  DTPOSTED: DateResponse
  TRNAMT: string
  FITID: string
  CHECKNUM: string
  MEMO: string
}

export type BankTransferList = {
  DTSTART: DateResponse
  DTEND: DateResponse
  STRTTRN: STRTTRN[]
}

type LedGerBal = {
  BALAMT: string
  DTASOF: DateResponse
}

type FINANCIAL_INSTITUTION = {
  ORG: string
  FID: string
}

export type OfxStructure = {
  OFX: {
    SIGNONMSGSRSV1: {
      SONRS: {
        STATUS: Status
        DTSERVER: DateResponse
        LANGUAGE: string
        DTACCTUP: DateResponse
        FI: FINANCIAL_INSTITUTION
      }
    }
    BANKMSGSRSV1: {
      STMTTRNRS: {
        TRNUID: number
        STATUS: Status
        STMTRS: {
          CURDEF: string
          BANKACCTFROM: BankAccountFrom
          BANKTRANLIST: BankTransferList
          LEDGERBAL: LedGerBal
          MKTGINFO: string
        }
      }
    }
  }
}

export type OfxResponse = OFXMetaData & OfxStructure
