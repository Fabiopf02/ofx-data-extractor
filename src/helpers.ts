import { END_TEXT_BANK_TRANSFER, START_TEXT_BANK_TRANSFER } from './config'
import { STRTTRN as STRTTRNType } from './types'

function separatePartsOfDate(date: string) {
  const year = date.slice(0, 4)
  const month = date.slice(4, 6)
  const day = date.slice(6, 8)
  const hour = date.slice(8, 10)
  const minutes = date.slice(10, 12)
  const seconds = date.slice(12, 14)
  const [offset, timezone] = date
    .slice(14)
    .replace('[', '')
    .replace(']', '')
    .split(':')
  return {
    yyyy: year,
    yy: year.slice(2),
    y: year,
    MM: month,
    M: month,
    dd: day,
    d: day,
    hh: hour,
    h: hour,
    mm: minutes,
    m: minutes,
    ss: seconds,
    s: seconds,
    O: offset,
    TZ: timezone,
  }
}

export function formatDate(date: string, format: string) {
  const parts = separatePartsOfDate(date)
  let result = format
  for (const [key, value] of Object.entries(parts)) {
    result = result.replace(key, value)
  }
  return result
}

export function fixJsonProblems(content: string) {
  const result = content
    .replace(/(},})/g, '}}')
    .replace(/(}")/g, '},"')
    .replace(/(},])/g, '}]')
    .replace(/(,})/g, '}')
    .replace(/({")/g, '{\n"')
    .replace(/(})/g, '\n}')
    .replace(/(",")/g, '",\n"')
    .replace(/,\s*}/g, '\n}')
    .replace(/(,",)/, ',')
  return result.at(-1) === ',' ? result.slice(0, -1) : result
}

export function extractFinancialInstitutionTransactionId(fitid: string) {
  const dateText = fitid.slice(0, 12)
  const transactionCode = fitid.slice(12, 19)
  const protocol = fitid.slice(19)
  const resultObject = {
    date: dateText,
    transactionCode,
    protocol,
  }
  return JSON.stringify(resultObject)
}

export function isDateField(field: string) {
  return field.startsWith('DT')
}

export function trim(str: string) {
  return str.trim()
}

export function objectStartReplacer(param: string) {
  if (param === START_TEXT_BANK_TRANSFER) return param
  return param.replace(/[<]/g, '\n').replace(/[>]/g, ':{')
}

export function objectEndReplacer(param: string) {
  if (param === END_TEXT_BANK_TRANSFER) return `\n${param}`
  return '},'
}

export function sanitizeCurrency(value: string) {
  const comma = value.search(',')
  const point = value.search('.')
  if (comma > point) return value.replace(/[.]/g, '').replace(/[,]/g, '.')
  return value.replace(/[,]/g, '')
}

export function isValidNumberToConvert(field: string, value: string) {
  if (field.endsWith('ID') || field.endsWith('NUM')) return false
  return !isNaN(Number(value))
}

export function getBankTransferListText(ofxContent: string) {
  const startIndex = ofxContent.indexOf(START_TEXT_BANK_TRANSFER)
  const endIndex =
    ofxContent.lastIndexOf(END_TEXT_BANK_TRANSFER) +
    END_TEXT_BANK_TRANSFER.length
  const oldListText = ofxContent.substring(startIndex, endIndex)
  const startRgx = new RegExp(START_TEXT_BANK_TRANSFER, 'g')
  const endRgx = new RegExp(END_TEXT_BANK_TRANSFER, 'g')
  const newListText = `"STRTTRN":[${oldListText
    .replace(startRgx, '{')
    .replace(endRgx, '},')}]`
  return { oldListText, newListText }
}

export function convertMetaDataToObject(
  stringList: string[],
  nativeTypes: boolean,
) {
  const result: { [key: string]: any } = {}
  for (const line of stringList) {
    const [key, value] = line.split(':')
    const sanitizedKey = key.replace('\n', '')
    result[sanitizedKey] =
      nativeTypes && isValidNumberToConvert(key, value) ? Number(value) : value
  }
  return result
}

export function getTransactionsSummary(STRTTRN: STRTTRNType[]) {
  return STRTTRN.reduce(
    (prevValue, currValue) => {
      if (currValue.TRNTYPE.toLocaleLowerCase().startsWith('deb')) {
        prevValue.amountOfDebits++
        prevValue.debit += Number(currValue.TRNAMT.replace('-', ''))
        return prevValue
      }
      prevValue.amountOfCredits++
      prevValue.credit += Number(currValue.TRNAMT)
      return prevValue
    },
    { credit: 0, debit: 0, amountOfCredits: 0, amountOfDebits: 0 },
  )
}

export function bufferToString(data: Buffer) {
  return data.toString()
}

export async function fileFromPathToString(pathname: string) {
  const fileData: string = await new Promise((resolve, reject) => {
    import('fs').then(fs => {
      return fs.readFile(pathname, (err, data) => {
        if (err) reject(err)
        else resolve(data.toString())
      })
    })
  })
  return fileData
}

export async function blobToString(blob: Blob): Promise<string> {
  const data: string = await new Promise((resolve, reject) => {
    if (typeof window !== 'undefined' && window.FileReader) {
      const reader = new window.FileReader()
      reader.onload = event => resolve(event.target!.result as string)
      reader.onerror = event => reject(event.target!.error)
      reader.readAsText(blob)
    } else {
      reject(new Error('FileReader is not available in this environment.'))
    }
  })
  return data
}
