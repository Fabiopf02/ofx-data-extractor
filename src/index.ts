export { Ofx } from './main'
export { formatDate, fixJsonProblems } from './helpers'
export type { OfxResponse, OFXMetaData, OfxStructure, BankTransferList } from './types'
