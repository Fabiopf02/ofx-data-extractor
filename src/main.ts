import { ELEMENT_CLOSURE_REGEX, ELEMENT_OPENING_REGEX } from './config'
import {
  blobToString,
  bufferToString,
  convertMetaDataToObject,
  extractFinancialInstitutionTransactionId,
  fileFromPathToString,
  fixJsonProblems,
  formatDate,
  getBankTransferListText,
  getTransactionsSummary,
  isDateField,
  isValidNumberToConvert,
  objectEndReplacer,
  objectStartReplacer,
  trim,
} from './helpers'
import {
  OfxConfig,
  OFXMetaData,
  OfxResponse,
  OfxStructure,
  STRTTRN,
} from './types'

export class Ofx {
  private _data: string
  private _config: OfxConfig = {}

  constructor(data: string, config?: OfxConfig) {
    if (config) this._config = config
    this._data = data
  }

  static fromBuffer(data: Buffer) {
    return new Ofx(bufferToString(data))
  }

  static async fromFilePath(pathname: string) {
    const data = await fileFromPathToString(pathname)
    return new Ofx(data)
  }

  static async fromBlob(blob: Blob): Promise<Ofx> {
    const data = await blobToString(blob)
    return new Ofx(data)
  }

  config(config: OfxConfig) {
    this._config = config
    return this
  }

  getHeaders(): OFXMetaData {
    const [metaDataString] = this._data.split('<OFX>')
    const metaDataList = metaDataString.split('\n')
    const validate = (line: string) => !!line.trim().length
    const validatedMetaDataList = metaDataList.filter(validate)
    return convertMetaDataToObject(
      validatedMetaDataList,
      !!this._config.nativeTypes,
    ) as OFXMetaData
  }

  getBankTransferList(): STRTTRN[] {
    const { newListText } = getBankTransferListText(this.getPartialJsonData())

    const list = newListText.slice(10)
    const fixedList = fixJsonProblems(list)
    return JSON.parse(fixedList)
  }

  private getPartialJsonData() {
    const [_, ofxContentText] = this._data.split('<OFX>')
    const ofxContent = '<OFX>' + ofxContentText
    const { sanitize } = this
    /**
     * Use replace first for closing tag so there is no conflict in `objectStartReplacer`
     */
    return ofxContent
      .replace(ELEMENT_CLOSURE_REGEX, objectEndReplacer)
      .replace(ELEMENT_OPENING_REGEX, objectStartReplacer)
      .split('\n')
      .map(trim)
      .filter(Boolean)
      .map(sanitize, this)
      .join('')
  }

  private sanitizeValue(field: string, value: string) {
    let fieldValue = value.replace(/[{},]/g, '')
    if (isDateField(field)) {
      fieldValue = this.configDate(fieldValue)
    } else if (field === 'FITID') {
      return this.configFinancialInstitutionTransactionId(fieldValue)
    } else if (
      this._config.nativeTypes &&
      isValidNumberToConvert(field, fieldValue)
    ) {
      return `${fieldValue},`
    }
    return `"${fieldValue}",`
  }

  private sanitize(line: string) {
    let sanitizedLine = line
    const field = sanitizedLine.slice(0, sanitizedLine.indexOf(':'))
    const sanitizeValue = this.sanitizeValue
    const replacer = (matched: string) =>
      sanitizeValue.call(this, field, matched)
    if (line.match(/{(\w|\W)+/)) {
      sanitizedLine = sanitizedLine.replace(/({(\w|\W)+)$/, replacer)
    }
    const matchedProperty = sanitizedLine.search(/(^\w+:)/)
    if (matchedProperty < 0) return sanitizedLine

    return sanitizedLine.replace(field, `"${field}"`)
  }

  private configFinancialInstitutionTransactionId(fitString: string) {
    const { fitId } = this._config
    if (fitId === 'separated')
      return extractFinancialInstitutionTransactionId(fitString)
    return `"${fitString}",`
  }

  private configDate(dateString: string) {
    const { formatDate: format } = this._config
    if (format) return formatDate(dateString, format)
    return formatDate(dateString, 'y-M-d')
  }

  getTransactionsSummary() {
    const jsonData = this.getContent()
    const { DTEND, DTSTART, STRTTRN } =
      jsonData.OFX.BANKMSGSRSV1.STMTTRNRS.STMTRS.BANKTRANLIST
    const summary = getTransactionsSummary(STRTTRN)
    return {
      dateStart: DTSTART,
      dateEnd: DTEND,
      ...summary,
    }
  }

  getContent(): OfxStructure {
    const ofxText = this.getPartialJsonData()
    const { newListText, oldListText } = getBankTransferListText(ofxText)
    const result = ofxText.replace(oldListText, newListText)
    return JSON.parse(`{${fixJsonProblems(result)}}`)
  }

  toJson(): OfxResponse {
    const ofxMetaDataResult = this.getHeaders()
    const ofxContentResult = this.getContent()
    const result = { ...ofxMetaDataResult, ...ofxContentResult }
    return result
  }
}
