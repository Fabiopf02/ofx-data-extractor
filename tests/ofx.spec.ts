import { Ofx } from '../src/index'
import path from 'path'
import fs from 'fs'

describe('Tests in the Node.js environment', () => {
  const file = fs.readFileSync(path.resolve(__dirname, 'example.ofx'))
  const ofx = Ofx.fromBuffer(file)

  test('It should read Buffer content', async () => {
    const file = fs.readFileSync(path.resolve(__dirname, 'example.ofx'))
    const ofx = Ofx.fromBuffer(file)
    const headers = ofx.getHeaders()
    expect(headers.OFXHEADER).toBe('100')
    expect(headers.CHARSET).toBe('1252')
    expect(headers.ENCODING).toBe('UTF-8')
    expect(headers.VERSION).toBe('102')
  })

  test.concurrent('It should return the correct amount of transactions', () => {
    expect(ofx.getBankTransferList()).toHaveLength(18)
  })

  test.concurrent('Should correctly return transaction summary', () => {
    const summary = ofx.getTransactionsSummary()
    expect(summary.credit).toBe(669.6)
    expect(summary.debit.toFixed(1)).toBe('34.1')
    expect(summary.amountOfCredits).toBe(9)
    expect(summary.amountOfDebits).toBe(9)
    expect(summary.dateStart).toBe('2018-01-30')
    expect(summary.dateEnd).toBe('2018-04-29')
  })

  test.concurrent('It should be equal to the snapshot', () => {
    const jsonData = ofx.toJson()
    expect(jsonData).toMatchSnapshot()
  })
  test.concurrent('It should be equal to the snapshot /2', () => {
    expect(ofx.getContent()).toMatchSnapshot()
  })

  test.concurrent('Should separate the FITID', () => {
    const toCompare = ofx.getBankTransferList()[0]
    const result = ofx.config({ fitId: 'separated' }).getBankTransferList()[0]
      .FITID as any
    expect(toCompare.FITID).toBe(
      result.date + result.transactionCode + result.protocol,
    )
  })

  test.concurrent('Should format the date correctly (d/M/y)', () => {
    const date = ofx
      .config({ formatDate: 'd/M/y' })
      .getBankTransferList()[0].DTPOSTED
    expect(date).toBe('09/03/2018')
  })

  test.concurrent('It should correctly return the offset and timezone', () => {
    const transfer = ofx.config({ nativeTypes: true }).getBankTransferList()[0]
    expect(typeof transfer.TRNAMT).toBe('number')
  })
})
