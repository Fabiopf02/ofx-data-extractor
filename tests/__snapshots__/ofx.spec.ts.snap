// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
{
  "<?OFX OFXHEADER="200" VERSION="211" SECURITY="NONE" OLDFILEUID="NONE" NEWFILEUID="NONE"?>": undefined,
  "CHARSET": "1252",
  "COMPRESSION": "NONE",
  "DATA": "OFXSGML",
  "ENCODING": "UTF-8",
  "NEWFILEUID": "NONE<?xml version="1.0" encoding="US-ASCII"?>",
  "OFX": {
    "BANKMSGSRSV1": {
      "STMTTRNRS": {
        "STATUS": {
          "CODE": "0",
          "SEVERITY": "INFO",
        },
        "STMTRS": {
          "BANKACCTFROM": {
            "ACCTID": "1459950-11",
            "ACCTTYPE": "CHECKING",
            "BANKID": "364",
          },
          "BANKTRANLIST": {
            "DTEND": "2018-04-29",
            "DTSTART": "2018-01-30",
            "STRTTRN": [
              {
                "CHECKNUM": "000000061553574",
                "DTPOSTED": "2018-03-09",
                "FITID": "2018030607232002046000000061553574",
                "MEMO": "Repasse pagamento: 17223405 de XXXXXXXX",
                "REFNUM": "2002046000000017223405",
                "TRNAMT": "74.40",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000061553576",
                "DTPOSTED": "2018-03-09",
                "FITID": "2018030607231001046000000061553576",
                "MEMO": "Tarifa repasse: 17223405 de XXXXXXXX",
                "TRNAMT": "-3.34",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000062976602",
                "DTPOSTED": "2018-03-20",
                "FITID": "2018031703312002046000000062976602",
                "MEMO": "Repasse pagamento: 22612776 de YYYYYYYYYYY",
                "REFNUM": "2002046000000022612776",
                "TRNAMT": "74.40",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000062976603",
                "DTPOSTED": "2018-03-20",
                "FITID": "2018031703311001046000000062976603",
                "MEMO": "Tarifa repasse: 22612776 de YYYYYYYYYYY",
                "TRNAMT": "-3.34",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000063575983",
                "DTPOSTED": "2018-03-30",
                "FITID": "2018032704142002046000000063575983",
                "MEMO": "Repasse pagamento: 28015595 de Fulano Aguiar",
                "REFNUM": "2002046000000028015595",
                "TRNAMT": "74.40",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000063575986",
                "DTPOSTED": "2018-03-30",
                "FITID": "2018032704141001046000000063575986",
                "MEMO": "Tarifa repasse: 28015595 de Fulano Aguiar",
                "TRNAMT": "-3.34",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000064109592",
                "DTPOSTED": "2018-04-06",
                "FITID": "2018040307262002046000000064109592",
                "MEMO": "Repasse pagamento: 28108174 de Ciclano da Silva",
                "REFNUM": "2002046000000028108174",
                "TRNAMT": "65.10",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000064109593",
                "DTPOSTED": "2018-04-06",
                "FITID": "2018040307261001046000000064109593",
                "MEMO": "Tarifa repasse: 28108174 de Ciclano da Silva",
                "TRNAMT": "-3.34",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000065551500",
                "DTPOSTED": "2018-04-16",
                "FITID": "2018041303342002046000000065551500",
                "MEMO": "Repasse pagamento: 29384993 de ZZZZZZZZZZ",
                "REFNUM": "2002046000000029384993",
                "TRNAMT": "74.40",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000065551501",
                "DTPOSTED": "2018-04-16",
                "FITID": "2018041303341001046000000065551501",
                "MEMO": "Tarifa repasse: 29384993 de ZZZZZZZZZZ",
                "TRNAMT": "-3.34",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000065642576",
                "DTPOSTED": "2018-04-17",
                "FITID": "2018041403362002046000000065642576",
                "MEMO": "Repasse pagamento: 29447637 de OOOOOOOOO",
                "REFNUM": "2002046000000029447637",
                "TRNAMT": "55.80",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000065642577",
                "DTPOSTED": "2018-04-17",
                "FITID": "2018041403361001046000000065642577",
                "MEMO": "Tarifa repasse: 29447637 de OOOOOOOOO",
                "TRNAMT": "-3.34",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000065694866",
                "DTPOSTED": "2018-04-28",
                "FITID": "2018041421312002047000000065694866",
                "MEMO": "Repasse pagamento: 29517140 de Beltrano Diniz",
                "REFNUM": "2002047000000029517140",
                "TRNAMT": "120.90",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000065694867",
                "DTPOSTED": "2018-04-28",
                "FITID": "2018041421311001047000000065694867",
                "MEMO": "Tarifa repasse: 29517140 de Beltrano Diniz",
                "TRNAMT": "-7.23",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000065696376",
                "DTPOSTED": "2018-04-29",
                "FITID": "2018041517302002047000000065696376",
                "MEMO": "Repasse pagamento: 29533821 de HA",
                "REFNUM": "2002047000000029533821",
                "TRNAMT": "55.80",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000065696377",
                "DTPOSTED": "2018-04-29",
                "FITID": "2018041517301001047000000065696377",
                "MEMO": "Tarifa repasse: 29533821 de HA",
                "TRNAMT": "-3.49",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000066643667",
                "DTPOSTED": "2018-04-29",
                "FITID": "2018042606102002046000000066643667",
                "MEMO": "Repasse pagamento: 30830691 de Duque",
                "REFNUM": "2002046000000030830691",
                "TRNAMT": "74.40",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000066643670",
                "DTPOSTED": "2018-04-29",
                "FITID": "2018042606101001046000000066643670",
                "MEMO": "Tarifa repasse: 30830691 de Duque",
                "TRNAMT": "-3.34",
                "TRNTYPE": "DEBIT",
              },
            ],
          },
          "CURDEF": "BRL",
          "LEDGERBAL": {
            "BALAMT": "635.50",
            "DTASOF": "2018-04-29",
          },
        },
        "TRNUID": "0",
      },
    },
    "CREDITCARDMSGSRSV1": {
      "CCSTMTTRNRS": {
        "CCSTMTRS": {
          "BANKTRANLIST": {
            "DTEND": "2005-08-31",
            "DTSTART": "2005-08-01",
            "STRTTRN": [
              {
                "DTPOSTED": "2005-08-11",
                "FITID": "219867",
                "NAME": "Interest Charge",
                "TRNAMT": "-23.00",
                "TRNTYPE": "INT",
              },
              {
                "DTPOSTED": "2005-08-11",
                "FITID": "219868",
                "NAME": "Payment - Thank You",
                "TRNAMT": "350.00",
                "TRNTYPE": "CREDIT",
              },
            ],
          },
          "CCACCTFROM": {
            "ACCTID": "123412341234",
          },
          "CURDEF": "USD",
          "LEDGERBAL": {
            "BALAMT": "-562.00",
            "DTASOF": "2005-08-31",
          },
        },
        "STATUS": {
          "CODE": "0",
          "SEVERITY": "INFO",
        },
        "TRNUID": "0",
      },
    },
    "SIGNONMSGSRSV1": {
      "SONRS": {
        "DTSERVER": "2005-08-31",
        "LANGUAGE": "ENG",
        "STATUS": {
          "CODE": "0",
          "SEVERITY": "INFO",
        },
      },
    },
  },
  "OFXHEADER": "100",
  "OLDFILEUID": "NONE",
  "SECURITY": "NONE",
  "VERSION": "102",
}
`;

exports[` 2`] = `
{
  "OFX": {
    "BANKMSGSRSV1": {
      "STMTTRNRS": {
        "STATUS": {
          "CODE": "0",
          "SEVERITY": "INFO",
        },
        "STMTRS": {
          "BANKACCTFROM": {
            "ACCTID": "1459950-11",
            "ACCTTYPE": "CHECKING",
            "BANKID": "364",
          },
          "BANKTRANLIST": {
            "DTEND": "2018-04-29",
            "DTSTART": "2018-01-30",
            "STRTTRN": [
              {
                "CHECKNUM": "000000061553574",
                "DTPOSTED": "2018-03-09",
                "FITID": "2018030607232002046000000061553574",
                "MEMO": "Repasse pagamento: 17223405 de XXXXXXXX",
                "REFNUM": "2002046000000017223405",
                "TRNAMT": "74.40",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000061553576",
                "DTPOSTED": "2018-03-09",
                "FITID": "2018030607231001046000000061553576",
                "MEMO": "Tarifa repasse: 17223405 de XXXXXXXX",
                "TRNAMT": "-3.34",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000062976602",
                "DTPOSTED": "2018-03-20",
                "FITID": "2018031703312002046000000062976602",
                "MEMO": "Repasse pagamento: 22612776 de YYYYYYYYYYY",
                "REFNUM": "2002046000000022612776",
                "TRNAMT": "74.40",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000062976603",
                "DTPOSTED": "2018-03-20",
                "FITID": "2018031703311001046000000062976603",
                "MEMO": "Tarifa repasse: 22612776 de YYYYYYYYYYY",
                "TRNAMT": "-3.34",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000063575983",
                "DTPOSTED": "2018-03-30",
                "FITID": "2018032704142002046000000063575983",
                "MEMO": "Repasse pagamento: 28015595 de Fulano Aguiar",
                "REFNUM": "2002046000000028015595",
                "TRNAMT": "74.40",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000063575986",
                "DTPOSTED": "2018-03-30",
                "FITID": "2018032704141001046000000063575986",
                "MEMO": "Tarifa repasse: 28015595 de Fulano Aguiar",
                "TRNAMT": "-3.34",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000064109592",
                "DTPOSTED": "2018-04-06",
                "FITID": "2018040307262002046000000064109592",
                "MEMO": "Repasse pagamento: 28108174 de Ciclano da Silva",
                "REFNUM": "2002046000000028108174",
                "TRNAMT": "65.10",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000064109593",
                "DTPOSTED": "2018-04-06",
                "FITID": "2018040307261001046000000064109593",
                "MEMO": "Tarifa repasse: 28108174 de Ciclano da Silva",
                "TRNAMT": "-3.34",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000065551500",
                "DTPOSTED": "2018-04-16",
                "FITID": "2018041303342002046000000065551500",
                "MEMO": "Repasse pagamento: 29384993 de ZZZZZZZZZZ",
                "REFNUM": "2002046000000029384993",
                "TRNAMT": "74.40",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000065551501",
                "DTPOSTED": "2018-04-16",
                "FITID": "2018041303341001046000000065551501",
                "MEMO": "Tarifa repasse: 29384993 de ZZZZZZZZZZ",
                "TRNAMT": "-3.34",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000065642576",
                "DTPOSTED": "2018-04-17",
                "FITID": "2018041403362002046000000065642576",
                "MEMO": "Repasse pagamento: 29447637 de OOOOOOOOO",
                "REFNUM": "2002046000000029447637",
                "TRNAMT": "55.80",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000065642577",
                "DTPOSTED": "2018-04-17",
                "FITID": "2018041403361001046000000065642577",
                "MEMO": "Tarifa repasse: 29447637 de OOOOOOOOO",
                "TRNAMT": "-3.34",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000065694866",
                "DTPOSTED": "2018-04-28",
                "FITID": "2018041421312002047000000065694866",
                "MEMO": "Repasse pagamento: 29517140 de Beltrano Diniz",
                "REFNUM": "2002047000000029517140",
                "TRNAMT": "120.90",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000065694867",
                "DTPOSTED": "2018-04-28",
                "FITID": "2018041421311001047000000065694867",
                "MEMO": "Tarifa repasse: 29517140 de Beltrano Diniz",
                "TRNAMT": "-7.23",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000065696376",
                "DTPOSTED": "2018-04-29",
                "FITID": "2018041517302002047000000065696376",
                "MEMO": "Repasse pagamento: 29533821 de HA",
                "REFNUM": "2002047000000029533821",
                "TRNAMT": "55.80",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000065696377",
                "DTPOSTED": "2018-04-29",
                "FITID": "2018041517301001047000000065696377",
                "MEMO": "Tarifa repasse: 29533821 de HA",
                "TRNAMT": "-3.49",
                "TRNTYPE": "DEBIT",
              },
              {
                "CHECKNUM": "000000066643667",
                "DTPOSTED": "2018-04-29",
                "FITID": "2018042606102002046000000066643667",
                "MEMO": "Repasse pagamento: 30830691 de Duque",
                "REFNUM": "2002046000000030830691",
                "TRNAMT": "74.40",
                "TRNTYPE": "CREDIT",
              },
              {
                "CHECKNUM": "000000066643670",
                "DTPOSTED": "2018-04-29",
                "FITID": "2018042606101001046000000066643670",
                "MEMO": "Tarifa repasse: 30830691 de Duque",
                "TRNAMT": "-3.34",
                "TRNTYPE": "DEBIT",
              },
            ],
          },
          "CURDEF": "BRL",
          "LEDGERBAL": {
            "BALAMT": "635.50",
            "DTASOF": "2018-04-29",
          },
        },
        "TRNUID": "0",
      },
    },
    "CREDITCARDMSGSRSV1": {
      "CCSTMTTRNRS": {
        "CCSTMTRS": {
          "BANKTRANLIST": {
            "DTEND": "2005-08-31",
            "DTSTART": "2005-08-01",
            "STRTTRN": [
              {
                "DTPOSTED": "2005-08-11",
                "FITID": "219867",
                "NAME": "Interest Charge",
                "TRNAMT": "-23.00",
                "TRNTYPE": "INT",
              },
              {
                "DTPOSTED": "2005-08-11",
                "FITID": "219868",
                "NAME": "Payment - Thank You",
                "TRNAMT": "350.00",
                "TRNTYPE": "CREDIT",
              },
            ],
          },
          "CCACCTFROM": {
            "ACCTID": "123412341234",
          },
          "CURDEF": "USD",
          "LEDGERBAL": {
            "BALAMT": "-562.00",
            "DTASOF": "2005-08-31",
          },
        },
        "STATUS": {
          "CODE": "0",
          "SEVERITY": "INFO",
        },
        "TRNUID": "0",
      },
    },
    "SIGNONMSGSRSV1": {
      "SONRS": {
        "DTSERVER": "2005-08-31",
        "LANGUAGE": "ENG",
        "STATUS": {
          "CODE": "0",
          "SEVERITY": "INFO",
        },
      },
    },
  },
}
`;
